src/cashing.rs

```diff
@@ -0,0 +1,141 @@
+//! Basic casher struct which stores a closure and a hashmap.
+//! The hasmap stores key value pairs representing previous
+//! function calls.
+//!
+//! When the Casher function is run, it first does a lookup
+//! to see if the value has already been calculated. If it has,
+//! it returns that value. If it hasn't, it calculates the value,
+//! adds it to the hashmap, and returns it.
+
+use std::collections::HashMap;
+
+/// The Casher struct (Memoization) stores a function and a Hashmap.
+/// The HashMap keeps track of previous input and output for the function so
+/// that it only ever has to be called once. Use for expensive functions.
+pub struct Casher<F, U, V>
+where
+    F: Fn(U) -> V,
+    U: std::cmp::Eq + std::hash::Hash + Copy,
+    V: Copy,
+{
+    calculation: F,
+    values: HashMap<U, V>,
+}
+
+impl<F, U, V> Casher<F, U, V>
+where
+    F: Fn(U) -> V,
+    U: std::cmp::Eq + std::hash::Hash + Copy,
+    V: Copy,
+{
+    /// Constuctor for the Casher
+    /// # Examples
+    /// ```
+    /// # use contest_algorithms::cashing::Casher;
+    /// let mut squared = Casher::new(|n: u32| n*n);
+    /// ```
+    pub fn new(calculation: F) -> Casher<F, U, V> {
+        Casher {
+            calculation,
+            values: HashMap::new(),
+        }
+    }
+
+    /// Performs a lookup into the hashmap to see if the value has already
+    /// been calculated. If it has, returns the value. If it has not,
+    /// calls the function, stores the value, then returns the value
+    /// # Examples
+    /// ```
+    /// # use contest_algorithms::cashing::Casher;
+    /// let mut squared = Casher::new(|n: u32| n*n);
+    ///
+    /// // This is where we call the function
+    /// let sixteen = squared.on(4);
+    /// ```
+    pub fn on(&mut self, arg: U) -> V {
+        // Clippy doesn't like that this isn't lazy but I'm fine with it.
+        *self.values.entry(arg).or_insert((self.calculation)(arg))
+    }
+
+    /// Calls the function without performing a lookup and replaces
+    /// the old calculation with the new one, then returns the value
+    ///
+    /// # Use Case
+    /// If you're wondering, this is for if some sort of "state" has changed
+    /// underneath you, so your same function call with the same input
+    /// might now have different output. For instance, if part of your function
+    /// reads from a file and
+    /// you think the contents of that file have changed even though the name
+    /// has not.
+    pub fn call_and_replace(&mut self, arg: U) -> V {
```

Comment:
Hm this might be useful for a small number of users. In the name of conciseness, I'm not sure yet if I'd keep this feature in the long-term, since simple problems like those in competitions shouldn't have this kind of state-dependence. Anyway, for now I'll keep it however you decide.

Reply:
Even if it doesn't make sense for use in competition, I think this function makes a lot of sense for the structure. I can see it being used with rand in clever ways as well.